pyads/pyads_ex.py

```diff
@@ -149,6 +150,87 @@ def adsAddRoute(net_id, ip_address):
+    import struct
+    from contextlib import closing
+    # ALL SENT STRINGS MUST BE NULL TERMINATED
+    adding_host_name = adding_host_name + '\0' if adding_host_name else socket.gethostname() + '\0'
+    added_net_id = added_net_id if added_net_id else sending_net_id
+    route_name = route_name + '\0' if route_name else adding_host_name
+
+    username = username + '\0'
+    password = password + '\0'
+
+    # The head of the UDP AMS packet containing host routing information
+    data_header = struct.pack('>12s', b'\x03\x66\x14\x71\x00\x00\x00\x00\x06\x00\x00\x00')
+    data_header += bytes(map(int, sending_net_id.split('.')))		# Sending net ID
+    data_header += struct.pack('<H', PORT_SYSTEMSERVICE)			# Internal communication port
+    data_header += struct.pack('>2s', b'\x05\x00')					# Write command
+    data_header += struct.pack('>4s', b'\x00\x00\x0c\x00')			# Block of unknown
+    data_header += struct.pack('<H', len(adding_host_name))			# Length of sender host name
+    data_header += adding_host_name.encode('utf-8')					# Sender host name
+    data_header += struct.pack('>2s', b'\x07\x00')					# Block of unknown
+
+
+    actual_data = struct.pack('<H', 6)								# Byte length of AMS ID (always 6)
+    actual_data += bytes(map(int, added_net_id.split('.')))			# Net ID being added to the PLC
+    actual_data += struct.pack('>2s', b'\x0d\x00')					# Block of unknown (maybe encryption?)
+    actual_data += struct.pack('<H', len(username))					# Length of the user name field
+    actual_data += username.encode('utf-8')							# PLC Username
+    actual_data += struct.pack('>2s', b'\x02\x00')					# Block of unknown
+    actual_data += struct.pack('<H', len(password))					# Length of password field
+    actual_data += password.encode('utf-8')							# PLC Password
+    actual_data += struct.pack('>2s', b'\x05\x00')					# Block of unknown
+    actual_data += struct.pack('<H', len(route_name))				# Length of route name
+    actual_data += route_name.encode('utf-8')						# Name of route being added to the PLC
+
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_DGRAM)) as sock: # UDP
+        # Listen on 55189 for the response from the PLC
+        sock.bind(('', 55189))
+
+        # Send our data to 48899 on the PLC
+        sock.sendto(data_header + actual_data, (ip_address, 48899))
+
+        # Response should come in in less than .5 seconds, but wait longer to account for slow communications
+        sock.settimeout(5)
+        # Allow TimeoutError to be raised so user can handle it how they please
+
+        # Keep looping until we get a response from our PLC
+        addr = [0]
+        while addr[0] != ip_address:
+            data, addr = sock.recvfrom(32) # PLC response is 32 bytes long
+
+    rcvd_packet_header = data[0:12]							# AMS Packet header, seems to define communication type
+    # If the last byte in the header is 0x80, then this is a response to our request
+    if rcvd_packet_header[-1] == 0x80:
```

Comment:
If `rcvd_packet_header[-1] != 0x80` then `rcvd_is_password_correct` will never be defined. It'll raise `UnboundLocalError` then.

Reply:
I'm not sure what it should do here. If the packet header doesn't end in 0x80 but it did come from the PLC on the port we were listening to it would mean we got a message we shouldn't have. Maybe intentionally raising an exception would be best?

Reply:
I can't speak for the repository maintainer, but I think raising a `RuntimeError("Unexpected response from PLC")` would not be unreasonable...